Add Octave HG version, use with purePackages.octave

asppsa opened the pull request:
###### Motivation for this change

Octave's embedding API has some issues in the current version (4.2.x) which causes `purePackages.octave` to segfault. Presumably this is an issue for other packages that use the embedding API also, although I am not really across this.    The issue is fixed in Octave's Mercurial repo, so I have worked on adding an HG version of Octave to nixpkgs.

This set of patches does the following:

- Adds a derivation for building Octave from Mercurial, and adds a listing for this to `all-packages.nix`.  This also required adding the `epstool` package.
- Changes `purePackages.octave` to use the mercurial version of Octave
- Changes the platforms to `none` for both the above packages so that they don't get built by Hydra.  I wasn't sure about this, so if you think it should be built on Hydra, I can change it.


###### Things done

I tried to test using sandboxing, but ran into a lot of errors with shebangs in packages that are needed to build (automake, etc.).  Not sure if it was something I was doing wrong, or if these packages are at fault.  Otherwise I did the bulk of my testing using `nix-shell --pure` on Archlinux.

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

asppsa committed:
epstool: add package

asppsa committed:
octaveHg: add package

asppsa committed:
purePackages.octave: Use octaveHg

Pure's Octave package compiles but crashes with Octave 4.2 due to an
upstream bug which is fixed in mercurial.

asppsa commented:
Sorry, there's something wrong with the epstool derivation.  I'll try to fix and then push to this branch again

7c6f434c commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, gfortran, readline, ncurses, perl, flex, texinfo, qhull
```
I think current default build does build info files — do you mean that it doesn't need `texinfo` for building them?

asppsa replied:
Texinfo is not required at build time because the source archive already has compiled versions of all the documentation (and also other machine-independent stuff, like scripts).  This can be seen by comparing the contents of `doc/interpreter` for the existing octave package and the proposed octaveHg one, e.g.,

~~~ bash
nix-shell -A octave '<nixpkgs>'
unpackPhase
ls octave-4.2.1/doc/interpreter/
~~~

For the `octave` package, you will see all the info, PDF, etc. files there already.

However, I think I've made an mistake, as texinfo *is* required, but at runtime rather than build time, as the `makeinfo` command is used by `help`.  Actually, this seems to be broken with current normal Octave already:

~~~
$ nix-shell --pure -p octave --run "octave --eval 'help sparse'" '<nixpkgs>'
sh: makeinfo: command not found
warning: help: Texinfo formatting filter exited abnormally; raw Texinfo source of help text follows...
'sparse' is a built-in function from the file libinterp/corefcn/sparse.cc



Additional help for built-in functions and operators is
available in the online version of the manual.  Use the command
'doc <topic>' to search the manual index.

Help and information about Octave is also available on the WWW
at http://www.octave.org and via the help@octave.org
mailing list.
~~~

`nix-shell --pure -p octave texinfo --run "octave --eval 'help sparse'" '<nixpkgs>'` works though. 

I'll rework this PR to make texinfo a runtime dep of Octave.

asppsa committed:
octave: add runtime texinfo dependency

makeinfo (provided by the texinfo package) is used by the "help"
command in Octave to display info about functions, etc.  By default,
Octave looks for "makeinfo" in the PATH, rather than specifying the
location of the executable.  This results in Nix not being aware that
makeinfo is required at runtime (so unless you happen have makeinfo
available from your path, "help" won't work).

This patch fixes that by setting the path to makeinfo in Octave,
thereby creating a runtime dependency on texinfo.

asppsa commented:
Hey @7c6f434c, so I've modified the PR now adding a patch to Octave that creates an explicit runtime dep on texinfo.  I also made a couple of changes to the octaveHg package: some reformatting, and also moved the bootstrap to `prePatch` instead of `preConfigure`.  Retested with `nox-review`

7c6f434c commented:
Not sure I care about bootstrap location unless there is a specific problem fixed by that. Intuitively it is more of a `preConfigure` thing, but your point about relative patches is also reasonable. Doesn't matter much, though…
